apps/service_providers/llm_service/runnables.py

```diff
@@ -205,16 +209,17 @@ def _populate_memory(self):
         messages = compress_chat_history(self.session.chat, model, self.experiment.max_token_limit)
         self.memory.chat_memory.messages = messages
 
-    def _save_message_to_history(self, message: str, type_: ChatMessageType):
+    def _save_message_to_history(self, message: str, type_: ChatMessageType, add_experiment_tag: bool = False):
         chat_message = ChatMessage.objects.create(
             chat=self.session.chat,
             message_type=type_.value,
             content=message,
         )
-        experiment_route = ExperimentRoute.objects.filter(team=self.session.team, child=self.experiment.id).first()
-        if not Tag.objects.filter(name=experiment_route.keyword, team=self.session.team).exists() and experiment_route:
-            chat_message.tags.create(team=self.session.team, name=experiment_route.keyword, is_system_tag=True)
-        chat_message.add_tags([experiment_route.keyword], team=self.session.team, added_by=None)
+        if add_experiment_tag:
+            exp_route = ExperimentRoute.objects.filter(team=self.session.team, child=self.experiment.id).first()
```

Comment:
Ooh, careful here. The first `ExperimentRoute` here might not be the correct one, since there can be multiple experiment routes in this team with this experiment as the child. We need to pass in the "parent" experiment as well to inform this query

Reply:
We could pass in the tag value directly since it's available in the caller

Reply:
[2fa9de3](https://github.com/dimagi/open-chat-studio/pull/440/commits/2fa9de34d6530c4357be1c2124bb0810ce16eded)